Object detect module

warmspringwinds opened the pull request:
I marked up the future module.
Changed the `mb_lbp` function to work faster.

vighneshbirodkar commented on skimage/future/__init__.py:
```diff
@@ -5,6 +5,6 @@
+from . import graph, objdetect
 
-__all__ = ['graph']
+__all__ = ['graph', 'objdetect']
```
I'd call this module `detection`. Any thoughts from others ?

JDWarner replied:
Definitely start with `detect`. I'd be fine with `detect`, `detect_object`, and probably others.

jni replied:
I agree with @JDWarner, go with `detect`. It mirrors most of our API, with some well-justified exceptions (`filters`) and others not so well justified, that we might consider updating for 1.0: `segment`, `restore`, `view`.

JDWarner replied:
Yes, agreed @jni.

vighneshbirodkar commented:
@warmspringwinds This is shaping up really well. Somethings I'd like to see addressed. Some of the old-timers ( if I may say so :smile: ) can help me here

1. I am not a fan of `cpdef` at all.
2. Are we sure we want to include OpenCV's trained XML ? Given that we would replace it with our own later on.
3. All XML files should do in `data` and be references by `data.xml.faces()` or something similar. The way you do it in the test is ugly
4. If we follow the same XML format as OpenCV should that get a mention ?
5. Clean up some rough edges, there are places where I can see commented code.

warmspringwinds commented:
1. Sure. We can make three functions instead of it. If @JDWarner and @jni will help us to find the best solution, it would be great.
2. I see no problem here. This is a very good file, I tried it and it worked really good. The thing is that I still not getting how OpenCV does the training under the scenes. They made it differently from how it's written in the paper. So it may be the case that we will have to create two different loaders for our file and for the OpenCV file. I am for the support of OpenCV files because it has really performant training script and a lot of ready to use files and people might want to use them.
3. Yeah, that might be a better idea. I will do it.
4. Yes, sure. When I am more a less done I will mention this in documentation.
5. Oh, yeah. Forgot to remove it. Thanks.

vighneshbirodkar commented:
@warmspringwinds 
Regarding 2. I am OK with having 2 files, but we should have only one implementation.

warmspringwinds commented:
@vighneshbirodkar 
We will have only one implementation of training, yes.
But we may support two types of files.

vighneshbirodkar commented:
@warmspringwinds I mean, we should never have 2 implantations of MB LBP detection.

jni commented on skimage/future/objdetect/cascade.pyx:
```diff
@@ -0,0 +1,225 @@
+            Stage current_stage
+            MBLBPStump current_stump
+            MBLBP current_feature
+
```
Don't put more than one blank line in a row within methods/functions.

jni commented:
@vighneshbirodkar @warmspringwinds sorry I don't quite understand the three functions discussion. Our general approach has been to have a cython inner function for performance critical bits, and a pure python wrapper. Is that an option here?

jni commented:
I also don't know anything about the xml but pre-trained models are in general a very good idea imho. Python in general is a "batteries included" language, so users would be right to expect such convenience. And I rarely favour the creation of a new format, so if something exists that we can parse easily enough, use that. 

I hope that helps a bit, sorry if I missed the point but I don't have time right now to delve into the history of this! Thanks both of you for your efforts!

warmspringwinds commented:
@jni 
Sorry, I will elaborate:
We need [the function](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/_texture.pyx#L278) to be `cdef`, but if we make it so, we won't be able to use it [here](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/texture.py#L300).

The other way to solve it is to put the function-wrapper in `_texture` file and make the first function `cdef`. But in this case the documentation for the function-wrapper won't be correctly generated, because the arguments of a function can't be read from the `pyx` file. So, I mean that in order for the function documentation to be correctly generated, it has to be in `py` file.

So, we came up with an ugly solution: to use tree functions: `cdef` and `def` in `pyx` where `def` wraps the `cdef` function and another function in `py` file that wraps `def` from `pyx` file.

JDWarner commented:
This is exactly the use case cpdef was designed for. What are the downsides of letting Cython handle the dual definition for us again, @vighneshbirodkar? There is a clear argument here for simplicity/conciseness?

jni commented on skimage/future/objdetect/setup.py:
```diff
@@ -0,0 +1,30 @@
+    from numpy.distutils.core import setup
+    setup(maintainer='scikit-image Developers',
+          maintainer_email='scikit-image@googlegroups.com',
+          description='Graph-based Image-processing Algorithms',
```
This clearly needs updating.

jni commented on skimage/future/objdetect/cascade.pyx:
```diff
@@ -0,0 +1,225 @@
+
+        return True
+
+    def load_xml(self, filename, eps=1e-5):
```
This function is enormous. There's a rule of thumb that functions should not be more than 25 lines long. This isn't a hard rule, but there are good reasons for it.

Candidates for abstraction are every point where you commented a chunk of code. Replace these chunks with well-named functions.

This comment applies to other parts of this PR, too.

jni commented:
@vighneshbirodkar @warmspringwinds I'm afraid I agree with @JDWarner. @vighneshbirodkar, your [link](http://ask.sagemath.org/question/8902/any-reason-not-to-use-cpdef-for-every-cython-function/) specifies the performance tradeoff when subclassing a function, but I don't think it would ever apply here. The answer shows when you wouldn't want to use `cpdef`, but there is no argument that you should *never* use `cpdef`. As @JDWarner says, this is exactly the case where you would want to use it.

At the very least, if we want to use three functions, we would want to see benchmarks showing a clear time advantage to the three function solution.

vighneshbirodkar commented:
@warmspringwinds I agree that `cpdef` function has no speed downsides in this case, but Cython does not allow `cpdef` function to release the GIL. Given how we plan to use `prange` later, a function acquiring the GIL can turn out to be a huge bottleneck.

jni commented:
@vighneshbirodkar @warmspringwinds I didn't know about the nogil restriction by cpdef. That would indeed be a compelling reason not to use it.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,303 @@
+
+        return scale_factors
+
+    def detect_single_scale(self, float[:, ::1] int_img, float scale, int step_ratio=1, int amount_of_threads=4):
```
Don't harcode, use `omp_get_max_threads`
See http://www.plutospin.com/files/OpenMP_reference.pdf and http://docs.cython.org/src/userguide/parallelism.html#using-openmp-functions

On my system this evaluates to 4.

warmspringwinds replied:
@vighneshbirodkar 
Yes, on my system it's `4` too. But there is one interesting thing.
When I set the amount of threads to `10` manually, it works faster.

vighneshbirodkar replied:
What processor do you have ?

warmspringwinds replied:
`Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz`

```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 69
Stepping:              1
CPU MHz:               759.000
BogoMIPS:              4589.61
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              3072K
NUMA node0 CPU(s):     0-3
```

warmspringwinds replied:
So far, when I test parallel scale computation it's much slower then the parallel image computation. Even though I am not doing any result saving in the scale parallelization part, and the parallel image computation is benchmarked with saving.

vighneshbirodkar replied:
Can you push that code ?

warmspringwinds replied:
@vighneshbirodkar Hm, that is really interesting. I just made a more thorough testing and the thing is: the parallel image computation is faster when the amount of threads that I specify is low(`<5`), but when I do `>5` it becomes faster. I just looked at my processor: I have 4 cores and the max amount of threads for each core is `2`. So I guess I have `8` threads maximum.

Yes, I will push it soon

vighneshbirodkar replied:
We should try to get this number somehow. I have two cores and 2 threads per core.

warmspringwinds replied:
@vighneshbirodkar 
wow, just played with the `schedule` parameter. Set it to `dynamic`, started to work faster even with low amount of threads. I guess I will read more about this thing.

warmspringwinds replied:
@vighneshbirodkar this occurs because when we spawn threads for each scale they have unequal amount of work (the first scales are the hardest ones) and, therefore, we need to allow them to take up more resources when they need it.

So, we stick with scale threading and `dynamic`. I hope that my tests were right :)

vighneshbirodkar replied:
@warmspringwinds I don't think your PC can run 8 threads.You have 4 CPUs, not 4 cores.
http://ark.intel.com/products/75459/Intel-Core-i5-4200U-Processor-3M-Cache-up-to-2_60-GHz

warmspringwinds replied:
@vighneshbirodkar Yes, my bad, I made a mistake.

I have a strange behavior in my code.
I did the scale loop, but now every time I get different results and I can't figure out why.
I enclosed my files here, they can be compiled directly:
https://gist.github.com/warmspringwinds/8a1348038df52586905c

Please, have a look if you have time.

vighneshbirodkar replied:
@warmspringwinds Check gitter

vighneshbirodkar commented:
@warmspringwinds This is fantastic, I'm sure with some optimizations, we will have a really fast parallel face detector.

warmspringwinds commented:
Just faced the issue https://github.com/scikit-image/scikit-image/issues/867#issuecomment-33630037
it turned out `appveyor` also has it.

vighneshbirodkar commented on bento.info:
```diff
@@ -34,7 +34,7 @@ Library:
         skimage, skimage.color, skimage.data, skimage.draw, skimage.exposure,
         skimage.feature, skimage.filters, skimage.future, skimage.future.graph,
-        skimage.graph, skimage.io,
+        skimage.graph, skimage.io, skimage.future.detect_obj
```
@stefanv Your thoughts on the module name ?

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+
+        # Check if memory was allocated.
+        if not (features_carr and stumps_carr and stages_carr and LUTs_carr):
+            raise MemoryError()
```
Include a description of the error as an argument.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+        self._load_xml(xml_file, eps)
+
+
+    cdef int classify(self, float[:, ::1] int_img, Py_ssize_t row, Py_ssize_t col, float scale) nogil:
```
Can be changed to `bint`

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+        current_scale = 1
+        current_size = np.array((self.window_height, self.window_width))
+
+        while (current_size <= max_size).all():
```
Loop can be replaced
```python
    min_size = np.array(min_size, dtype=np.float32)
    max_size = np.array(max_size, dtype=np.float32)

    row_power_max = math.log(max_size[0]/current_size[0], scale_step)
    col_power_max = math.log(max_size[1]/current_size[1], scale_step)

    row_power_min = math.log(min_size[0]/current_size[0], scale_step)
    col_power_min = math.log(min_size[1]/current_size[1], scale_step)

    mn = max(row_power_min, col_power_min)
    mx = min(row_power_max, col_power_max)

    powers = np.arange(mn, mx)

    multipliers = np.power(scale_step, powers)
    return multipliers
```

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+            current_col = 0
+
+            while current_row < max_row:
+                while current_col < max_col:
```
Move 
`current_col = 0` here, for readability.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+import xml.etree.ElementTree as ET
+from ...feature._texture cimport _multiblock_lbp
+
+
```
A one line explanation of each struct here might be helpful.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+
+        Returns
+        -------
+        output : list of dicts
```
We could modify the original struct to be more verbose like `row` and `column`

vighneshbirodkar replied:
@stefanv @jni @JDWarner 
What's the best thing to return here ?

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,488 @@
+            C-contiguous integral image of the input image.
+        """
+
+        img = rgb2gray(img)
```
We cannot handle 2D arrays right now if we blindly use this call.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,612 @@
+
+    for current_detection in range(1, detections_amount):
+
+        best_score = 0.5
```
This should be taken as a parameter.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,612 @@
+            clusters[best_cluster_number].c += detections[current_detection].c
+            clusters[best_cluster_number].width += detections[current_detection].width
+            clusters[best_cluster_number].height += detections[current_detection].height
+            clusters_scores[best_cluster_number] += 1
```
A cluster struct will make things more readable.

vighneshbirodkar commented:
@warmspringwinds Ping ?

blink1073 commented:
@warmspringwinds, FYI, if you rebase on master, further Travis tests on this PR will be much faster.

warmspringwinds commented:
@blink1073 Thanks. I didn't know this.

blink1073 commented:
We just fixed it five minutes ago :smile:

warmspringwinds commented:
@vighneshbirodkar I included the cluster struct as you suggested. Now it looks more readable :)

vighneshbirodkar commented:
@warmspringwinds Have you run the python 3 tests locally ?

warmspringwinds commented:
@vighneshbirodkar No. But I am pretty sure this test will pass.

vighneshbirodkar commented on skimage/future/detect_obj/tests/test_cascade.py:
```diff
@@ -0,0 +1,29 @@
+                                               min_size=(24, 24),
+                                               max_size=(123, 123))
+
+        assert len(detected) == 2, 'Two faces on the image.'
```
Where is the 2nd face in the astronaut image ?

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,836 @@
+
+        # Check if memory was allocated.
+        if not (features_carr and stumps_carr and stages_carr and LUTs_carr):
+            raise MemoryError("Can't allocate memory.")
```
I would say something like "Failed to allocate memory while parsing XML".

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,836 @@
+        return True
+
+    def _get_valid_scale_factors(self, min_size, max_size, scale_step):
+        """Get the valid scale multipliers for the original window size
```
This short doc string needs to be one line only according to the docstring conventions.

vighneshbirodkar commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,836 @@
+                height = <Py_ssize_t>(current_feature.height * scale)
+
+
+                lbp_code = _multiblock_lbp(int_img,
```
Although PEP8 compatible, needless newlines should be avoided.

vighneshbirodkar commented:
@stefanv I think me and Daniil are pretty satisfied with the code now. We(mostly Daniil) have made it as fast as we could. Can you please give your thoughts and comment on the overall structure ? Also comment on if you think creation of a new module is justified or not.

JDWarner commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,836 @@
+from skimage._shared.transform cimport integrate
+
+# Use our own implementation of the function instead of libc.math.
+# Otherwise, the compilation breaks on some compilers due to Cython
```
"Otherwise, the compilation breaks on Windows due to a CPython bug." I'm fairly sure this isn't Cython's fault.

stefanv replied:
I'd also double check this

stefanv commented on skimage/data/__init__.py:
```diff
@@ -11,6 +11,7 @@
 from .. import data_dir
 from ..io import imread, use_plugin
 from ._binary_blobs import binary_blobs
+import xml
```
I would not group data by their formats, but rather by their purposes.  So if you want, use `skimage.data.detect.*`.

stefanv commented on skimage/data/lbpcascade_frontalface_opencv.xml:
```diff
@@ -0,0 +1,1505 @@
+<?xml version="1.0"?>
```
The name states where this file comes from, but also add a comment into the file itself showing its origin (URL, description, etc.)

stefanv commented on skimage/data/xml.py:
```diff
@@ -0,0 +1,10 @@
+
+def face_cascade_detector():
+    """
+    Returns the filepath to the trained xml file.
```
This description is not detailed enough to understand the content of the file

stefanv commented on skimage/feature/_texture.pyx:
```diff
@@ -340,8 +331,8 @@ def _multiblock_lbp(float[:, ::1] int_image,
 
-        current_rect_r = central_rect_r + r_offsets[element_num]
-        current_rect_c = central_rect_c + c_offsets[element_num]
+        current_rect_r = central_rect_r + mlbp_r_offsets[element_num]*height
```
PEP8

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+# added, the sum of row, col positions and width and heights are stored
+# with the count of rectangles that belong to this cluster. This way,
+# we don't have to store all the rectangles information as array
+# and the average can be easily computed in a constant time.
```
the average of what?

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+
+
+# Struct for storing multi-block binary pattern position.
+cdef struct MBLBP:
```
This looks like the same struct as above--can you make one struct, e.g., rectangle?

warmspringwinds replied:
@stefanv Yes, we can. But the code this way is more readable.
I mean otherwise it can be confusing.
In case of MBLBP is mean the top-left rectangle coordinates and in case
of detection I just mean the rectangle that covers a detection region.

stefanv replied:
Fair enough.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    Py_ssize_t height
+
+
+# Struct for storing a stump of classifying cascade. It has the index to the
```
This description should be clarified

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+
+# Struct for storing a stump of classifying cascade. It has the index to the
+# look-up table which is stored in Cascade class. Depending on the value of
+# the feature after its evaluation `left` or `right` value is returned which
```
What does left or right refer to here?

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+cdef vector[Detection] _group_detections(vector[Detection] detections,
+                                         float intersection_score_threshold=0.5,
+                                         int min_neighbour_amount=4):
+    """Groups similar detection into single detection and eliminates weak
```
Group similar detections into a single detection and eliminate weak (non-overlapping) detections.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    """Groups similar detection into single detection and eliminates weak
+    detections that have small amount of intersecting detection.
+
+    The function assumes that true detections are supposed to have a certain
```
We assume that a true detection is characterized by a high number of overlapping detections.  Such detections are isolated and gathered into one cluster.  The average of each cluster is returned.

^^ Not sure what it means to "average a cluster"--so explain that as well.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+        Detection mean_detection
+        vector[DetectionsCluster] clusters
+        vector[int] clusters_scores
+        Py_ssize_t clusters_amount
```
clusters_amount is a strange variable naming--what does it signify?  Perhaps nr_of_clusters or cluster_overlap_nr etc. depending on what it means.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    Parameters
+    ----------
+    clusters : vector[DetectionsCluster]
+        Array of rectnagles clusters.
```
Spelling

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    clusters : vector[DetectionsCluster]
+        Array of rectnagles clusters.
+    count_threshold : int
+        The threshold amount of rectangles that is used.
```
``amount`` is reserved for uncountables such as "sugar", "water",etc.  For things you can count use "number of".

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+cdef float rect_intersection_area(Detection rect_a, Detection rect_b):
+    """Computes the intersection area of two rectangles.
+
+    The area where rectangles intersect.
```
Remove this line?

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    rect_a : Detection
+        Struct of the first rectnagle.
+    rect_a : Detection
+        Struct of the second rectnagle.
```
Typo

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    Parameters
+    ----------
+    rect_a : Detection
+        Struct of the first rectnagle.
```
Typo

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    Parameters
+    ----------
+    rect_a : Detection
+        Struct of the first rectnagle.
```
Typo

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    rect_a : Detection
+        Struct of the first rectnagle.
+    rect_a : Detection
+        Struct of the second rectnagle.
```
Typo

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+    Returns
+    -------
+    result : float
+        The intersection score. The number in the interval [0, 1].
```
Double back-ticks around ``` ``[0, 1]`` ```.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+
+
+cdef class Cascade:
+    """Class for cascade classifiers that are used for object detection."""
```
Give a better explanation of what a cascade is here.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+            stumps_amount += weak_classifiers_amount
+
+        # Allocate memory for data.
+        features_carr = <MBLBP*>malloc(features_amount*sizeof(MBLBP))
```
PEP8

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+                           intersection_score_threshold=0.5):
+        """Search for the object on multiple scales of input image.
+
+        The function takes the input image, the scale factor by which the
```
You can remove this paragraph (duplicates the parameter description below), and rather expand on what the function does.

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+        Returns
+        -------
+        output : list of dicts
+            Dict have form {'r': int, 'c': int, 'width': int, 'height': int},
```
List keys in a list?

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+        scale_factors = self._get_valid_scale_factors(min_size, max_size, scale_factor)
+        amount_of_scales = scale_factors.shape[0]
+
+        # Initialize lock to enable thread-safe writes to the array
```
I'm unsure about the use of OpenMP and of its limitations.  We'll have to talk about the scikit-learn / dipy teams, or other experts, on whether this is feasible at the moment.  Also, it is necessary to explicitly invoke these locks when Cython provides its own ``with parallel`` functionality?

vighneshbirodkar replied:
@stefanv Yes it is, `with` won't make a single `vector` thread-safe. The alternative is to use an array of vectors and no locks, but this solution is much cleaner.

dipy uses `prange` with some safety checks
https://github.com/nipy/dipy/blob/master/dipy/align/bundlemin.pyx

scikit-learn currently chooses not to use it
https://github.com/scikit-learn/scikit-learn/pull/4873

But I think we can explicitly document the limitations, and keep the implementation in `future`. I am specially tempted because of how exclusive detection at each scale is.

stefanv replied:
The question is, can we even get this compiled on all platforms?  E.g.,
can we use openmp under Windows?

stefanv commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,832 @@
+
+        # As the amount of work between the threads is not equal we use `dynamic`
+        # schedule which enables them to use computing power on demand.
+        for scale_number in prange(0, amount_of_scales, schedule='dynamic', nogil=True):
```
amount -> nr

stefanv commented on skimage/future/detect_obj/tests/test_cascade.py:
```diff
@@ -0,0 +1,29 @@
+                                               min_size=(60, 60),
+                                               max_size=(123, 123))
+
+        assert len(detected) == 1, 'One face should be detected.'
```
Also do a check on the position of the detected face.  And do this one at least two images!

vighneshbirodkar commented:
@warmspringwinds I think me and most others agree that the module should be called `detect` if at all we choose to have one

warmspringwinds commented:
@vighneshbirodkar Yeah, we can change it later. No problem.

@vighneshbirodkar, @stefanv Yes, we can make it `feature` but then we have one problem:
this framework will be able to detect not only faces and for each of those we will have to create a separate feature module.

Btw, it works really good on Lena image:
![lena_detect](https://cloud.githubusercontent.com/assets/2501383/8823142/dc27a444-306a-11e5-936d-e782a0f19449.png)

I have spent time reading about opencv's file format and now I am sure that I will be able to make
such a script after training. Also I created a dataset for training. I have spent quite a while tweaking OpenCV's file for augmenting training samples that add noise and slightly transforms images. It was a hell, but now it works.
I am starting with training and I think that we won't be able to use scikit-learn's function for this.
the gentle adaboost is simple and can be hardcoded, otherwise by making a call to scikit learn all the time will take forever to train. Haar-feature object detection takes week to train with parallel tweaks, so we' better speed up everything as much as possible.

jni commented on skimage/future/detect_obj/cascade.pyx:
```diff
@@ -0,0 +1,878 @@
+        return list(_group_detections(output, intersection_score_threshold,
+                                      min_neighbour_number))
+
+    def _load_xml(self, xml_file, eps=1e-5):
```
@warmspringwinds can you refactor this to use more functions? Every "paragraph" that you start with a comment explaining its purpose is a candidate for refactoring.

NASA [enforces](http://pixelscommander.com/wp-content/uploads/2014/12/P10.pdf) that no function should be longer than 60 lines, including variable declarations (one per line). I think we should stick to that — it allows code review having the full function in sight, rather than having to remember what came before.

This comment also applies to `detect_multi_scale`, above.

warmspringwinds replied:
@jni I will do it. Although this function is just reading an xml.

jni replied:
@warmspringwinds "is just reading an xml" != "will never need maintenance"

stefanv commented:
> I have spent time reading about opencv's file format and now I 
> am sure that I will be able to make such a script after 
> training. Also I created a dataset for training. I have spent 
> quite a while tweaking OpenCV's file for augmenting training 
> samples that add noise and slightly transforms images. It was a 
> hell, but now it works.  I am starting with training and I think 
> that we won't be able to use scikit-learn's function for this.

Can you please expand on what you wrote above?

warmspringwinds commented:
@stefanv I was not sure that I will be able to make the file with the same structure as in OpenCV after training (xml file with trained weights). Because I didn't understand how it work. Now I made it clear for myself.

I also created a dataset for training our own classifier using openCv createsamples script (it adds random transformations and noise).

stefanv commented:
There's a lot of detail here I'd like to understand.  Can you put the
detail in a blog post, please?

warmspringwinds commented:
@stefanv Basically you don't have to understand the first part. I just eliminated my confusions and now I am sure that training will work.

I will write a blog post about creating a dataset. Good idea.

vighneshbirodkar commented:
@stefanv Yes. We can compile on windows, else the appveyor build would fail. Although visual studio 2008 can't compile it, 2010 can.

vighneshbirodkar commented:
@warmspringwinds
I think all object detections can be considered features

stefanv commented:
In other words, we are limited to Python 3 for this functionality
(Python 2 package is compiled with 2008).

vighneshbirodkar commented:
@stefanv I don't think so, appveyor build is running fine for python 2.7

JDWarner commented:
@stefanv the AppVeyor build uses Python 2.7.

If I recall correctly, OpenMP under Clang in OSX has been the main issue with parallelism in the past. It seems significant strides have been made toward this support by [the clang-omp project](https://clang-omp.github.io/) since last I checked.

stefanv commented:
Let's ping @matthew-brett on this one.

stefanv commented:
We should also investigate whether we can parallelize via dask, since we're already releasing the GIL.

stefanv commented:
To profile C code (old way):

```
valgrind --tool=callgrind --callgrind-out-file=callgrind.log python sometask.py
```

Then visualize the result using kcachegrind.

To profile C code (new way):

http://docs.cython.org/src/tutorial/profiling_tutorial.html

warmspringwinds commented:
@stefanv @vighneshbirodkar 
I have updated the code to safely work with OpenMP in case the system doesn't have it like in the
dipy repository.

I have also added the gallery example.

I also checked the dask and unfortunately we can't use it because all of our code is written in
the cython already and dask wants everything to be written using pure numpy in order to work (like the dask's array that duplicates numpy function calls and then does computation in parallel on chunks of data).

stefanv commented:
Note that dask can be used simply as a scheduler--there is no need to use its numpy functionality, necessarily.  Anyway, we can leave this issue for later investigation.

stefanv commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+Face Detection
+==============
+
+This example shows how to detect faces on an image using object detection framework.
```
the object detection framework

stefanv commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+
+This example shows how to detect faces on an image using object detection framework.
+
+First, you will need an xml file from which the trained data can be read.
```
Why are we using XML files?  Let's convert the data to a more sensible format?

stefanv commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+The framework works with files trained using Multi-block Local Binary Patterns
+Features (See `MB-LBP <plot_multiblock_local_binary_pattern.html>`_) and
+Gentle Adaboost with attentional cascade. So, the detection framework will also
+work with `xml files from OpenCV <https://github.com/Itseez/opencv/tree/master/data/lbpcascades>`_.
```
Happy to support OpenCV .xml in addition.

stefanv commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+
+You should also take into account that false detections are inevitable and if you want
+to have a really precise detector, you will have to train it yourself using
+`OpenCV train cascade utility <http://docs.opencv.org/doc/user_guide/ug_traincascade.html>`_.
```
Can you update us about the latest state of the training in scikit-image?

warmspringwinds replied:
@stefanv it's about to be finished. I wrote a blog post where I wrote about the 
reason of delay.
http://warmspringwinds.github.io/gsoc/2015/08/23/google-summer-of-code-implementing-the-training-part-of-face-detection/

stefanv commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+from matplotlib import patches
+
+# Load the trained file from the module root.
+trained_file = data.detect.frontal_face_cascade_xml()
```
I'd remove the trailing `_xml`.

warmspringwinds replied:
@stefanv will do. Also seems better for me.

stefanv commented on skimage/data/detect.py:
```diff
@@ -0,0 +1,16 @@
+from .. import data_dir
+
+
+def frontal_face_cascade_xml():
```
Why not use this function to parse the XML as well?

warmspringwinds replied:
@stefanv I made it inside the original function because it takes time to load
and parse that file and it's a little bit faster in cython.
Taking into account that usually people will use this framework for detection only on one image, the loading should also be fast.

stefanv commented:
Can you please [measure Cython coverage](http://blog.behnel.de/posts/coverage-analysis-for-cython-modules.html) on this module?

warmspringwinds commented:
@stefanv @matthew-brett 

I made a new commit that is supposed to make compilation process complete successfully even on the compilers that don't support openmp.

I used an approach from dipy https://github.com/nipy/dipy/blob/f9c39fa243cdca0ea19baeda70be1a9417704c17/setup_helpers.py#L117

I ended up like this:
https://github.com/warmspringwinds/scikit-image/blob/820abe6c41ad656fafc67b259362971deb0436f3/skimage/future/detect/setup.py#L20

But I have one problem still: when I run ```python setup.py build_ext -i ``` locally in the detect module, it compiles everything correct (first it checks if headers are supported by current compiler and adds them if they are supported). But when I run ```python setup.py build_ext -i ``` from the root of the package, it doesn't do this.

Will appreciate any help here.

JDWarner commented:
First of all, thanks so much for sticking with this PR and looking into conditional OpenMP support @warmspringwinds! It looks like you're really close. Getting that running is hugely important, and will have large performance implications for Cython across the package.

Concerning the most recent Travis failure, it seems to be an issue with `print` being a function in Python 3. The recommended fix is to always call `print()` as a function, and add the line `from __future__ import print_function` at the top of the file. This makes Python 2.x behave like Python 3.x.

JDWarner commented on skimage/future/detect/setup.py:
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python
+
```
`from __future__ import print_function` goes here, before other imports.

JDWarner commented on skimage/future/detect/setup.py:
```diff
@@ -0,0 +1,94 @@
+    def build_extensions(self):
+        """ Hook into extension building to check compiler flags """
+
+        print '-----------------------'
```
Here's the errant `print` call Python 3.x doesn't like - change to `print('-----------------------')` after adding the above import and this should work (or... we'll graduate to a new, more exciting error ;)

warmspringwinds commented:
@JDWarner Sorry, I just made all my checks using python 2.
It should build now.

Anyways, I still have to same problem that I had before:
when I run setup from detect module the conditional inclusion of headers works, but it doesn't work when I call it from the root of the library.

Thank you.

warmspringwinds commented:
by doesn't work I mean that it compiles but the check for support of headers and their inclusion is not held.

warmspringwinds commented:
@vighneshbirodkar @JDWarner @stefanv 

Now the conditional adding of openmp headers and linking was added.
I had to modify the main setup file.

This way it checks if compiler supports headers and only then adds them into compilation process.
I still have to check it on mac os with clang. I will need help here.

And let's see if the travis works first.

JDWarner commented:
All lights are green!

warmspringwinds commented:
@JDWarner Yeah, we need to check if clang works.

@stefanv Could you please try to compile this pull request with clang your machine, or maybe someone else.

arokem commented:
Regarding compilation with OpenMP on clang (mac os x). It is possible, if you are using homebrew python or the python.org python (not Anaconda) and if clang is installed from homebrew with the following: 

    brew install clang-omp

And if `setup.py` conains the following: 

    os.environ['CC'] = '/usr/local/bin/clang-omp

or if the compiler is somehow otherwise linked there.

warmspringwinds commented:
@arokem Can you, please, check if my pull request compiles on your machine with clang?
(I mean without the openMP clang support).
I want to make the compilation possible on the machines without openMP support (want to check the conditional adding of headers).

arokem commented:
Seems to compile fine. I get the following error when I run the tests, but
maybe it's unrelated?

```

======================================================================

ERROR: Failure: AttributeError ('module' object has no attribute
'__pyx_capi__')

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/local/lib/python2.7/site-packages/nose/loader.py", line 418,
in loadTestsFromName

    addr.filename, addr.module)

  File "/usr/local/lib/python2.7/site-packages/nose/importer.py", line 47,
in importFromPath

    return self.importFromDir(dir_path, fqname)

  File "/usr/local/lib/python2.7/site-packages/nose/importer.py", line 94,
in importFromDir

    mod = load_module(part_fqname, fh, filename, desc)

  File "/Users/arokem/source/scikit-image/skimage/future/__init__.py", line
8, in <module>

    from . import graph, detect

  File
"/Users/arokem/source/scikit-image/skimage/future/detect/__init__.py", line
1, in <module>

    from .cascade import Cascade

  File "skimage/future/detect/cascade.pyx", line 1, in init
skimage.future.detect.cascade (skimage/future/detect/cascade.cpp:23779)

    # cython: cdivision=True

AttributeError: 'module' object has no attribute '__pyx_capi__'


----------------------------------------------------------------------

```



On Tue, Jan 5, 2016 at 4:52 PM, Daniil Pakhomov <notifications@github.com>
wrote:

> @arokem <https://github.com/arokem> Can you, please, check if my pull
> request compiles on your machine with clang?
> (I mean without the openMP clang support).
> I want to make the compilation possible on the machines without openMP
> support (want to check the conditional adding of headers).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/1570#issuecomment-169181526>
> .
>

warmspringwinds commented:
@arokem 
Thank you so much for you help!

Yeah, it's relevant.

Does it give you the same result when you use your trick with clang-omp for compiling?

stefanv commented:
Thanks, Ariel, for giving a hand here.

Daniil, I am away until Sunday, but happy to organize an OSX test account for you then.

warmspringwinds commented:
@stefanv That would be great!

OrkoHunter commented on doc/examples/plot_face_detection.py:
```diff
@@ -0,0 +1,93 @@
+# Initialize the detector cascade.
+detector = detect.Cascade(trained_file)
+
+img = data.lena()
```
We should make sure, `lena` doesn't go in again.

jni replied:
:+1: :100:

stefanv replied:
Daniil hasn't worked on this since his GSoC ended--I imagine he has been
quite busy setting up his next research position.  Daniil, do you have a
bit of time to round this one up?

warmspringwinds replied:
@stefanv Sure.

I have been mostly working first to make it compile on clang.
Is it still possible to get access to machine with clang?

jni commented:
I don't know the history of this but @warmspringwinds asked me to test this with CLang and I did and it works fine!

```console
(ana3)
jni@airbears2-10-142-107-176 Thu Jun 09 18:14
 ~/projects/scikit-image (objdetect)$ nosetests skimage/future/detect/tests/test_cascade.py
..
----------------------------------------------------------------------
Ran 2 tests in 2.984s

OK
```

warmspringwinds commented:
@JDWarner 
@stefanv 
@vighneshbirodkar 
@jni 

I have removed the lena image, updated documentation.
And here are some pictures :P

First image is an .xml file for frontal human face.
Second -- human profile human face.
third -- frontal cat face.

![skimage_faces](https://cloud.githubusercontent.com/assets/2501383/15994583/e3dc5bea-30d6-11e6-8977-25476660c9a1.png)
![skimage_profile](https://cloud.githubusercontent.com/assets/2501383/15994584/e3df8464-30d6-11e6-8c24-1a6e416e2cce.png)
![skimage_cats](https://cloud.githubusercontent.com/assets/2501383/15994585/e3e1c418-30d6-11e6-8cef-bcc41bf1d225.png)

stefanv commented:
:D​

RONNCC commented:
excited for this :D

jni commented:
@warmspringwinds huh? Since when do we host notebooks, let alone notebooks in the root dir?

The kitty face detector is awesome.

asnt commented:
Thank you for your work on this module.
Just want to report that it builds and runs nicely on my side. (Linux 64bit, Python3.5, OpenMP)

souravsingh commented:
@warmspringwinds Is the feature ready to be merged in scikit-image? I am really excited to have the feature in scikit-image.

jondo commented:
What's the state of this?

stefanv commented:
@warmspringwinds This may now be our longest living PR :)  Let's revisit:

1. Is this functionality still relevant for inclusion?
2. If so, are you willing to carry it the last mile, or should we find some more help?

glemaitre commented:
Just jumping in (I might need a cascade classifier somewhere).
I can give an hand if needed. Two questions:

* Would it be useful to accept Haar-like features as well as MB-LBP?
* Regarding clang-openmp support, what is the reason to not request `gcc` if people want to build from source. I assume that you are shipping wheel anyway?

jni commented:
@glemaitre as a heavy mac user, I would be upset if we required gcc for compilation. It's not trivial to get gcc working in macOS. (Actually even getting clang is kind of annoying.) I also like to support different compilers, which might have different optimisation characteristics.

glemaitre commented:
>  I also like to support different compilers, which might have different optimisation characteristics.

Good point.

stefanv commented:
Superceded by #3267.
